### ISSUE TITLE
prettyprint output with libcprintf

### ISSUE THREAD

**Issue by rountree** [LLNL/variorum#359]
Many of the examples output long tables of data that are suitable for machine reading but not so much for human reading.  An example is the fixed counters output (taken from octomore):

```
_FIXED_COUNTERS Host Thread InstRet UnhaltClkCycles UnhaltRefCycles
_FIXED_COUNTERS octomore 0 0 281471532535052 0
_FIXED_COUNTERS octomore 1 0 281458475422836 0
_FIXED_COUNTERS octomore 2 0 281462898555984 0
_FIXED_COUNTERS octomore 3 0 281457494630455 0
_FIXED_COUNTERS octomore 4 0 281467802590566 0
_FIXED_COUNTERS octomore 5 0 281463676113211 0
_FIXED_COUNTERS octomore 6 0 281455104202735 0
_FIXED_COUNTERS octomore 7 0 281462428128377 0
_FIXED_COUNTERS octomore 8 0 281470713199829 0
_FIXED_COUNTERS octomore 9 0 281470876233635 0
_FIXED_COUNTERS octomore 10 0 281470797754295 0
```

This can be made much more readable using the `printf()` minimum-width field.  Applying that looks like:

```
_FIXED_COUNTERS     Host Thread InstRet UnhaltClkCycles UnhaltRefCycles
_FIXED_COUNTERS octomore      0       0 281471532535052               0
_FIXED_COUNTERS octomore      1       0 281458475422836               0
_FIXED_COUNTERS octomore      2       0 281462898555984               0
_FIXED_COUNTERS octomore      3       0 281457494630455               0
_FIXED_COUNTERS octomore      4       0 281467802590566               0
_FIXED_COUNTERS octomore      5       0 281463192983359               0
_FIXED_COUNTERS octomore      6       0 281455104202735               0
_FIXED_COUNTERS octomore      7       0 281462428128377               0
_FIXED_COUNTERS octomore      8       0 281470713199829               0
_FIXED_COUNTERS octomore      9       0 281470876233635               0
_FIXED_COUNTERS octomore     10       0 281470752253162               0
```

Unfortunately, hard-coding minimum widths isn't feasible (one doesn't know how wide the hostname will be ahead of time, cycles can range from 0 to many, many digits).

I've created the `cprintf` library[1] to manage this kind of column printing.  `cprintf()` (and `cfprintf()`) work nearly exactly* like their regular counterparts, except that no printing occurs until `cflush()` is called.  At that point, the library parses all of the saved parameters from the previous `cprintf()` calls, determines the appropriate column widths, edits the format strings to add/modify those widths, and then prints as usual using the new format strings.  Here's the diff that resulted in the output above.

```
--- a/src/variorum/Intel/counters_features.c
+++ b/src/variorum/Intel/counters_features.c
@@ -10,6 +10,10 @@
 #include <sys/time.h>
 #include <unistd.h>
 
+#ifdef VARIORUM_CPRINTF
+#include <cprintf.h>
+#endif
+
 #include <clocks_features.h>
 #include <counters_features.h>
 #include <config_architecture.h>
@@ -243,8 +247,13 @@ void print_fixed_counter_data(FILE *writedest, off_t *msrs_fixed_ctrs)
 
     if (!init)
     {
+#ifdef VARIORUM_CPRINTF
+        cfprintf(writedest,
+                "%s %s %s %s %s %s\n", "_FIXED_COUNTERS", "Host", "Thread", "InstRet", "UnhaltClkCycles", "UnhaltRefCycles");
+#else
         fprintf(writedest,
                 "_FIXED_COUNTERS Host Thread InstRet UnhaltClkCycles UnhaltRefCycles\n");
+#endif
         init = 1;
     }
     variorum_get_topology(NULL, NULL, &nthreads);
@@ -254,9 +263,17 @@ void print_fixed_counter_data(FILE *writedest, off_t *msrs_fixed_ctrs)
     read_batch(FIXED_COUNTERS_DATA);
     for (i = 0; i < nthreads; i++)
     {
+#ifdef VARIORUM_CPRINTF
+        cfprintf(writedest, "%s %s %d %lu %lu %lu\n", "_FIXED_COUNTERS", hostname, i,
+                *c0->value[i], *c1->value[i], *c2->value[i]);
+#else
         fprintf(writedest, "_FIXED_COUNTERS %s %d %lu %lu %lu\n", hostname, i,
                 *c0->value[i], *c1->value[i], *c2->value[i]);
+#endif
     }
+#ifdef VARIORUM_CPRINTF
+    cflush();
+#endif
 }
```

While I think this is an interesting result, I do have a few concerns about moving this into mainline.

1. The `cprintf` code needs a lot of polishing and hardening, and we probably don't want to use it until it has undergone review and release.
2. The `cprintf` interface is sharp and brittle.  Modern compilers do a good job of detecting parameters/format mismatches.  All of that disappears if you use `cprintf`.  If there is an issue, it's likely garbage will be printed, rather than detecting the error condition.  That can lead to a lot of wasted time while someone tracks down why their counters are nonsense.
3. I'd like to support both machine-readable and human-readable formats (which might help in debugging).  Our interfaces would need to be modified to handle that option.

PR to follow.

* Asterisk and dollar notation are not yet supported.  You've probably never used either.  I certainly haven't.

[1] [cprintf repo](https://github.com/rountree/cprintf)

**Comment by sanmai-NL** [comment]
Alternatively, you could focus on machine-readable and interchangeable output (e.g., TOML or JSON) and pipe it through UI-like utilities. The more C string munging you do, the more security risk and performance efficiency degradation.

**Comment by slabasan** [comment]
Resolved by #468
